code/modules/modular_computers/NTNet/device_types/_network_device.dm

```diff
@@ -0,0 +1,180 @@
+/datum/extension/network_device
+	base_type = /datum/extension/network_device
+	expected_type = /atom/movable
+	flags = EXTENSION_FLAG_IMMEDIATE
+	var/network_id
+	var/key
+	var/address
+	var/network_tag
+	var/connection_type = NETWORK_CONNECTION_WIRELESS
+
+/datum/extension/network_device/New(datum/holder, n_id, n_key, c_type, autojoin = TRUE)
+	..()
+	network_id = n_id
+	key = n_key
+	if(c_type)
+		connection_type = c_type
+	address = uppertext(NETWORK_MAC)
+	var/obj/O = holder
+	network_tag = "[uppertext(replacetext(O.name, " ", "_"))]-[num2hex(sequential_id(type))]"
+	if(autojoin)
+		if(network_id)
+			connect()
+		else
+			connect_to_any()
+	
+/datum/extension/network_device/Destroy()
+	disconnect()
+	. = ..()
+
+/datum/extension/network_device/proc/connect()
+	var/datum/computer_network/net = GLOB.computer_networks[network_id]
+	if(!net)
+		return FALSE
+	return net.add_device(src)
+
+/datum/extension/network_device/proc/disconnect()
+	var/datum/computer_network/net = GLOB.computer_networks[network_id]
+	if(!net)
+		return FALSE
+	return net.remove_device(src)
+
+/datum/extension/network_device/proc/check_connection(specific_action)
+	var/datum/computer_network/net = GLOB.computer_networks[network_id]
+	if(!net)
+		return FALSE
+	if(!net.check_connection(src, specific_action) || !net.add_device(src))
+		return FALSE
+	return net.get_signal_strength(src)
+	
+/datum/extension/network_device/proc/get_signal_wordlevel()
+	var/datum/computer_network/network = get_network()
+	if(!network)
+		return "Not Connected"
+	var/signal_strength = network.get_signal_strength(src)
+	if(signal_strength <= 0)
+		return "Not Connected"
+	else if(signal_strength <= 6)
+		return "Low Signal"
+	else
+		return "High Signal"
+
+/datum/extension/network_device/proc/get_nearby_networks()
+	var/list/networks = list()
+	for(var/id in GLOB.computer_networks)
+		var/datum/computer_network/net = GLOB.computer_networks[id]
+		if(net.check_connection(src))
+			networks |= id
+	return networks
+
+/datum/extension/network_device/proc/is_banned()
+	var/datum/computer_network/net = get_network()
+	if(!net)
+		return FALSE
+	return (address in net.banned_nids)
+
+/datum/extension/network_device/proc/get_network()
+	if(check_connection())
+		return GLOB.computer_networks[network_id]
+
+/datum/extension/network_device/proc/add_log(text)
+	var/datum/computer_network/net = get_network()
+	if(!net)
+		return FALSE
+	return net.add_log(text, network_tag)
+
+/datum/extension/network_device/proc/connect_to_any()
```

Comment:
Connect_to_any shouldn't connect to a network if a key is set on that network, imo.

Reply:
it won't, it'll fail at the connect()